src/java/org/apache/cassandra/repair/RepairSession.java

```diff
@@ -185,62 +196,70 @@ public Collection<InetAddressAndPort> endpoints()
         return state.commonRange.endpoints;
     }
 
-    public void trackValidationCompletion(Pair<RepairJobDesc, InetAddressAndPort> key, ValidationTask task)
+    public synchronized void trackValidationCompletion(Pair<RepairJobDesc, InetAddressAndPort> key, ValidationTask task)
     {
+        if (terminated)
+        {
+            task.abort(new RuntimeException("Session terminated"));
+            return;
+        }
         validating.put(key, task);
     }
 
-    public void trackSyncCompletion(Pair<RepairJobDesc, SyncNodePair> key, CompletableRemoteSyncTask task)
+    public synchronized void trackSyncCompletion(Pair<RepairJobDesc, SyncNodePair> key, CompletableRemoteSyncTask task)
     {
+        if (terminated)
+            return;
         syncingTasks.put(key, task);
     }
 
     /**
      * Receive merkle tree response or failed response from {@code endpoint} for current repair job.
      *
      * @param desc repair job description
-     * @param endpoint endpoint that sent merkle tree
-     * @param trees calculated merkle trees, or null if validation failed
+     * @param message containing the merkle trees or an error
      */
-    public void validationComplete(RepairJobDesc desc, InetAddressAndPort endpoint, MerkleTrees trees)
+    public void validationComplete(RepairJobDesc desc, Message<ValidationResponse> message)
     {
+        InetAddressAndPort endpoint = message.from();
+        MerkleTrees trees = message.payload.trees;
         ValidationTask task = validating.remove(Pair.create(desc, endpoint));
+        // replies without a callback get dropped, so if in mixed mode this should be ignored
+        ctx.messaging().send(message.emptyResponse(), message.from());
```

Comment:
What happens if we don't send this response?

Reply:
>  do we need it for a local initiator

no, but we don't really know that to special case

> What happens if we don't send this response? 

Lets say the response is dropped, then the repair hangs forever...  since validation and sync can take a long time they don't use normal callbacks and just wait forever...